ring/spread_minimizing_token_generator.go

```diff
@@ -0,0 +1,279 @@
+package ring
+
+import (
+	"container/heap"
+	"fmt"
+	"math"
+	"regexp"
+	"sort"
+	"strconv"
+
+	"github.com/go-kit/log"
+	"github.com/go-kit/log/level"
+	"golang.org/x/exp/slices"
+)
+
+const totalTokensCount = 1 << 32
+
+var (
+	instanceIDRegex          = regexp.MustCompile(`^(.*)-(zone-[a-z])-(\d+)$`)
+	errorBadInstanceIDFormat = func(instanceID string) error {
+		return fmt.Errorf("unable to extract instance id from \"%s\"", instanceID)
+	}
+	errorZoneNotValid = func(zone string) error {
+		return fmt.Errorf("zone %s is not valid", zone)
+	}
+	errorMultipleOfZonesCount = func(zonesCount, optimalTokenOwnership uint32, token ringToken) error {
+		return fmt.Errorf("calculation of a new token between %d and %d with optimal token ownership %d was impossible: optimal token ownership must be a positive multiple of number of zones %d", token.prevToken, token.token, optimalTokenOwnership, zonesCount)
+	}
+	errorLowerAndUpperBoundModulo = func(zonesCount, optimalTokenOwnership uint32, token ringToken) error {
+		return fmt.Errorf("calculation of a new token between %d and %d with optimal token ownership %d was impossible: lower and upper bounds must be congruent modulo number of zones %d", token.prevToken, token.token, optimalTokenOwnership, zonesCount)
+	}
+	errorDistanceBetweenTokensNotBigEnough = func(optimalTokenOwnership, ownership int, token ringToken) error {
+		return fmt.Errorf("calculation of a new token between %d and %d with optimal token ownership %d was impossible: distance between lower and upper bound %d is not big enough", token.prevToken, token.token, optimalTokenOwnership, ownership)
+	}
+)
+
+type SpreadMinimizingConfig struct {
+	InstanceID        string
+	Zone              string
+	TokensPerInstance int
```

Comment:
As discussed: changing the number of tokens per instance will not really work for this strategy (because the ownership decisions will most likely lead to different set of tokens, especially when not powers of 2). 

So we should hard code this to 512, as that's the number we know works and we recommend to the users
Verify in GenerateTokens that we're requested that number of tokens, and panic otherwise.